### ISSUE TITLE
MSan miss on O1 optimization level (for CVE-2016-20014)

### ISSUE THREAD

**Issue by Teemperor** [llvm/llvm-project#55595]
*Note*: I'm aware that this is happening due to a (legal) optimization and this is not technically a bug in the optimizer (nor the sanitizer implementation really). This (and a few other bugs I'm filing) mostly serve as a basis for an upcoming RFC on how to fix these issues.

The code below is reading uninitialized memory but on -O1 (which is the msan recommended level) this uninitialized read (and the free call) gets optimized away before the MSan pass. Note that both GCC and Clang actually end up turning undef into 0 so free is never called (unless you're using O0 or another compiler).

```
#include <stdlib.h>

struct rep {
  int *x;
};

int main() {
  struct rep r;
  if (r.x)
    free(r.x);
}
```
https://godbolt.org/z/zP535rEzM

The code above is a 1-1 copy of CVE-2016-20014 but I found the same bug in a bunch of libraries and applications (WINE, X11, Bullet, some file system drivers, etc.).

Some more info about what happens:

SROA replaces the access with an `undef` and the last time we see the `undef` is before IPSCC (see the optimizer log below):

```llvm
*** IR Dump After EarlyCSEPass on main ***
; Function Attrs: mustprogress norecurse nounwind sspstrong uwtable
define dso_local i32 @main() #0 {
  br i1 undef, label %1, label %2

1:                                                ; preds = %0
  br label %2

2:                                                ; preds = %1, %0
  ret i32 0
}
*** IR Dump After IPSCCPPass on [module] ***
; Function Attrs: mustprogress norecurse nounwind sspstrong uwtable
define dso_local i32 @main() #0 {
  br label %1

1:                                                ; preds = %0
  ret i32 0
}
```

**Comment by fhahn** [comment]
This is an interesting issue and similar to a few others.  LLVM aggressively makes use of `undef` and happily assumes UB doesn't happen. This is at odds with sanitizers which try to catch (sometimes obvious) UB.

At the moment I don't really have any good ideas on how to prevent this when using sanitizers, as in the end this is up to each individual pass to decide.

**Comment by fhahn** [comment]
(This probably would benefit from a wider discussion on discourse)

**Comment by Teemperor** [comment]
> (This probably would benefit from a wider discussion on discourse)

I agree. I posted the RFC here: https://discourse.llvm.org/t/rfc-safe-optimizations-for-sanitizers/62729

**Comment by fhahn** [comment]
It might be good to link all related issue together in the issue tracker as well, as it seems like you are discovering quite a few :)
